### ISSUE TITLE
Errors when all players leave server with VSH2 running

### ISSUE THREAD

**Issue by Starblaster64** [Starblaster64/Vs-Saxton-Hale-2#28]
Getting several errors when all players leave a server while VSH2 is running. Lots of invalid client errors.

L 09/25/2016 - 12:55:11: [SM] Exception reported: Invalid client index 0
L 09/25/2016 - 12:55:11: [SM] Blaming: vsh2.smx()
L 09/25/2016 - 12:55:11: [SM] Call stack trace:
L 09/25/2016 - 12:55:11: [SM]   [0] ThrowError
L 09/25/2016 - 12:55:11: [SM]   [1] Line 40, C:\hlserver\tf\addons\sourcemod\scripting\include\morecolors.inc::CPrintToChat()
L 09/25/2016 - 12:55:11: [SM]   [2] Line 1367, modules/handler.sp::_SkipBossPanel()
L 09/25/2016 - 12:55:11: [SM]   [4] Call_Finish
L 09/25/2016 - 12:55:11: [SM]   [5] Line 639, modules/stocks.inc::DoThink()
L 09/25/2016 - 12:55:11: [SM] Exception reported: Invalid client index 0
L 09/25/2016 - 12:55:11: [SM] Blaming: vsh2.smx()
L 09/25/2016 - 12:55:11: [SM] Call stack trace:
L 09/25/2016 - 12:55:11: [SM]   [0] ThrowError
L 09/25/2016 - 12:55:11: [SM]   [1] Line 40, C:\hlserver\tf\addons\sourcemod\scripting\include\morecolors.inc::CPrintToChat()
L 09/25/2016 - 12:55:11: [SM]   [2] Line 1367, modules/handler.sp::_SkipBossPanel()
L 09/25/2016 - 12:55:11: [SM]   [4] Call_Finish
L 09/25/2016 - 12:55:11: [SM]   [5] Line 639, modules/stocks.inc::DoThink()
L 09/25/2016 - 12:55:25: [SM] Exception reported: Client index 0 is invalid
L 09/25/2016 - 12:55:25: [SM] Blaming: vsh2.smx()
L 09/25/2016 - 12:55:25: [SM] Call stack trace:
L 09/25/2016 - 12:55:25: [SM]   [0] AreClientCookiesCached
L 09/25/2016 - 12:55:25: [SM]   [1] Line 110, modules/base.sp::BaseFighter.iPresetType.get()
L 09/25/2016 - 12:55:25: [SM]   [2] Line 398, modules/gamemode.sp::VSHGameMode.GetBossType()
L 09/25/2016 - 12:55:25: [SM]   [3] Line 135, modules/events.sp::RoundStart()

**Comment by Starblaster64** [comment]
stocks.inc line numbers might be a little off, since this is on the branch I'm doing all the model/material stuff on, but I didn't touch anything related to what's in the errors.

**Comment by assyrianic** [comment]
yea it errors out because it's not catching the disconnecting player fast enough. Do you know a hook/event that fires when a player is about to leave?

**Comment by x3m77rus** [comment]
https://wiki.alliedmods.net/Generic_Source_Server_Events#player_disconnect
Is EventHookMode_Pre will be fast enough?

**Comment by CookiesIO** [comment]
In some cases for _SkipNextBoss() there's simply not enough players and so FindNextBoss doesn't return a valid a player. The loop looks for the next 3 players, but if there's less than that on the server, it wont find any and thus will try to send a chat message to client 0 (which is, as seen in errors logs, not good).

The [upnext[j]](https://github.com/Starblaster64/Vs-Saxton-Hale-2/blob/develop/addons/sourcemod/scripting/modules/handler.sp#L1366) there should be checked if it's a valid client before continuing on.

That block should probably be something like this
```
{
	upnext[j] = gamemode.FindNextBoss();
	if (IsValidClient(upnext[j].index))
	{
		upnext[j].bSetOnSpawn = true;
		if (!j)
			SkipBossPanelNotify(upnext[j].index);
		else CPrintToChat(upnext[j].index, "{olive}[VSH]{default} You are going to be a Boss soon! Type {olive}/halenext{default} to check/reset your queue points.");
	}
}
```

**Comment by Starblaster64** [comment]
@CookiesIO Last time I remember seeing you was your release of BFF.
Anyway, I actually already fixed what you pointed out when I noticed morecolors throwing errors about printing to bots, I'll push it along with the Goomba Stomp integration when I get around to it.

**Comment by CookiesIO** [comment]
@Starblaster64 Haha, yeah, that's some time ago. Actually I've returned to working on BFF, which we'll use for the server(s) my new community runs when it's ready, until then we're using VSH2.

Alright, neat. Reason I pointed it out is that we have run into quite the error spam after running it for a few days, I have a short list with proposed fixes (I got the error log thrown at me while I was sick, so didn't feel digging too deep and fix it myself at the time). Should I create issues for each of them, with the proposed fixes in them?
